donkeycar/parts/tub_v2.py

```diff
@@ -120,13 +124,48 @@ def run(self, *args):
         self.tub.write_record(record)
         return self.tub.manifest.current_index
 
+    def run_threaded(self, *args):
+        assert len(self.tub.inputs) == len(args), \
+            f'Expected {len(self.tub.inputs)} inputs but received {len(args)}'
+        record = dict(zip(self.tub.inputs, args))
+        # push the record into queue
+        self.record_queue.put(record)
+        self.mutex.acquire()
+        count = self.tub.manifest.current_index
+        self.mutex.release()
+        return count
+
+    def update(self):
+        self.on = True
+        while self.on or not self.record_queue.empty() > 0:
+            try:
+                # pop a record
+                record = self.record_queue.get()
+                # write the record to disk
+                self.mutex.acquire()
```

Comment:
I don't think you need the lock here as this runs in a single continuous thread.

Reply:
As I my previous comment said, `self.tub` is being accessed from two threads. On main thread, its `manifest.current_index` will be read, while in this thread, its `manifest.write_record()` will be called eventually. Those need to be taken care of. Of course it can be better here to implement a read/write lock. However due to the complexity in the whole datastore, I try to avoid to do so and a mutex is working fast enough for me.

Reply:
Who is using `manifest.current_index` from the main thread ? If its being used that way - we should break that contract and hide the `current_index` completely from the main thread. Otherwise we introduce a lot of other racy-ness.

Reply:
`run_thread()` is using `manifest.current_index` from the main thread, as it needs to return the number of elements in the manifest. This behaviour aligns with `run()` but since now we're multithreading this writer, it needs to be protected. It doesn't need to be hidden, returning a local copy of the number should be fine.